src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp

```diff
@@ -220,55 +320,64 @@ bool BallPlacementPlayFSM::retreatDone(const Update &event)
+std::pair<Angle, Point> BallPlacementPlayFSM::calculateWallPickOffLocation(
+    const Point &ball_pos, const Rectangle &field_boundary, const double max_dist)
 {
-    Angle kick_angle;
-    if (ball_pos.x() > field_lines.xMax())
+    Angle facing_angle;
+    Point backoff_point;
+    bool near_positive_y_boundary = (field_boundary.yMax() - ball_pos.y()) < max_dist;
+    bool near_negative_y_boundary = (ball_pos.y() - field_boundary.yMin()) < max_dist;
+    bool near_positive_x_boundary = (field_boundary.xMax() - ball_pos.x()) < max_dist;
+    bool near_negative_x_boundary = (ball_pos.x() - field_boundary.xMin()) < max_dist;
+    if (near_positive_y_boundary && near_positive_x_boundary)  // top right corner
     {
-        if (ball_pos.y() > 0)
-        {
-            kick_angle = Angle::fromDegrees(45);
-        }
-        else
-        {
-            kick_angle = Angle::fromDegrees(-45);
-        }
+        facing_angle  = Angle::fromDegrees(45);
+        backoff_point = field_boundary.posXPosYCorner() -
+                        Vector::createFromAngle(facing_angle)
+                            .normalize(BACK_AWAY_FROM_CORNER_EXTRA_M);
     }
-    else if (ball_pos.x() < field_lines.xMin())
+    else if (near_positive_y_boundary && near_negative_x_boundary)  // top left corner
     {
-        if (ball_pos.y() > 0)
-        {
-            kick_angle = Angle::fromDegrees(135);
-        }
-        else
-        {
-            kick_angle = Angle::fromDegrees(-135);
-        }
+        facing_angle  = Angle::fromDegrees(135);
+        backoff_point = field_boundary.negXPosYCorner() -
+                        Vector::createFromAngle(facing_angle)
+                            .normalize(BACK_AWAY_FROM_CORNER_EXTRA_M);
     }
-    else if (ball_pos.y() > field_lines.yMax())
+    else if (near_negative_y_boundary && near_positive_x_boundary)  // bottom right corner
     {
-        if (ball_pos.x() > 0)
-        {
-            kick_angle = Angle::fromDegrees(135);
-        }
-        else
-        {
-            kick_angle = Angle::fromDegrees(45);
-        }
+        facing_angle  = Angle::fromDegrees(-45);
+        backoff_point = field_boundary.posXNegYCorner() -
+                        Vector::createFromAngle(facing_angle)
+                            .normalize(BACK_AWAY_FROM_CORNER_EXTRA_M);
     }
-    else if (ball_pos.y() < field_lines.yMin())
+    else if (near_negative_y_boundary && near_negative_x_boundary)  // bottom left corner
     {
-        if (ball_pos.x() > 0)
-        {
-            kick_angle = Angle::fromDegrees(-135);
-        }
-        else
-        {
-            kick_angle = Angle::fromDegrees(-45);
-        }
+        facing_angle  = Angle::fromDegrees(-135);
+        backoff_point = field_boundary.negXNegYCorner() -
+                        Vector::createFromAngle(facing_angle)
+                            .normalize(BACK_AWAY_FROM_CORNER_EXTRA_M);
+    }
```

Comment:
we don't have notions of top, bottom, left and right in our documentation. This shouldn't be a part of the comments.